### PR TITLE
feat: amplitude add support for unset

### PR DESCRIPTION
**Pull request by anantjain45823**
## What are the changes introduced in this PR?

Resolves INT-950
Adding support for [Amplitude unset field ](https://www.docs.developers.amplitude.com/analytics/apis/http-v2-api/#keys-for-the-event-argument:~:text=exceed%2040%20layers.-,user_properties,-Optional.%20Object.%20A) through Rudderstack Integration Object

## Please explain the objectives of your changes below

Adding support for [unset fields for amplitude](https://www.docs.developers.amplitude.com/analytics/apis/identify-api/#required-parameters:~:text=%24-,unset,-Removes%20a%20property) integration through integrations object

### Type of change

**enhancement**

- New capabilities/behaviour - Now `user_properties` can be unset through sending `integrations.Amplitude.fieldsToUnset` array containing list of fields/nested fields to unset

If the pull request is a **new feature**,

### Any technical or performance related pointers to consider with the change?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### If the PR has changes in more than 10 files, please mention why the changes were not split into multiple PRs.

N/A

### If multiple linear tasks are associated with the PR changes, please elaborate on the reason:

N/A

<hr>

### Developer checklist

- [x] **No breaking changes are being introduced.**

- [x] Are all related docs linked with the PR?

- [x] Are all changes manually tested?

- [x] Does this change require any documentation changes?

- [x] Are relevant unit and component test-cases added?

### Reviewer checklist

- [x] Is the type of change in the PR title appropriate as per the changes?

- [x] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

File: src/v0/destinations/am/transform.js [hunk 1]
```diff
@@ -488,6 +498,7 @@ const responseBuilderSimple = (
 
   // 2. get campaign info (only present for JS sdk and http calls)
   const campaign = get(message, 'context.campaign') || {};
+  //
```
Comment (yashasvibajpai) [inline src/v0/destinations/am/transform.js:501]:
```suggestion
```

File: src/v0/destinations/am/transform.js [hunk 2]
```diff
@@ -453,6 +462,7 @@ const nonAliasResponsebuilder = (
     };
     respList.push(groupResponse);
   }
+
```
Comment (yashasvibajpai) [inline src/v0/destinations/am/transform.js:465]:
```suggestion
```

File: test/integrations/destinations/am/processor/data.ts [hunk 3]
```diff
@@ -791,6 +796,9 @@ export const data = [
                       insert_id: '84e26acc-56a5-4835-8233-591137fca468',
                       ip: '0.0.0.0',
                       user_properties: {
+                        $unset: {
+                          'testObj.unsetField1': '-',
```
Comment (yashasvibajpai) [inline test/integrations/destinations/am/processor/data.ts:800]: Does this reflect correctly over at amplitude, like remove the nested property inside testObj?

**Review by yashasvibajpai — CHANGES_REQUESTED** [review]
Very minor formatting changes and a query.

File: test/integrations/destinations/am/processor/data.ts (hunk 3 above)
Comment (anantjain45823) [reply test/integrations/destinations/am/processor/data.ts:800]: Yes thats already verified

File: src/v0/destinations/am/transform.js [hunk 4]
```diff
@@ -297,6 +297,15 @@ const identifyBuilder = (message, destination, rawPayload) => {
       }
     });
   }
+  // update identify call request with unset fields
+  const fieldsToUnset = get(message, 'integrations.Amplitude.fieldsToUnset');
```
Comment (koladilip) [inline src/v0/destinations/am/transform.js:301]: can you document the format of data here

Comment (koladilip) [inline src/v0/destinations/am/transform.js:301]: Add amplitude docs about how to treat these properties

Comment (koladilip) [inline src/v0/destinations/am/transform.js:301]: can we move whole thing to a separate function

Comment (koladilip) [reply src/v0/destinations/am/transform.js:301]: so that we can write unit tests

Comment (anantjain45823) [reply src/v0/destinations/am/transform.js:301]: Ok

Comment (anantjain45823) [reply src/v0/destinations/am/transform.js:301]: done

Comment (anantjain45823) [reply src/v0/destinations/am/transform.js:301]: done

Comment (anantjain45823) [reply src/v0/destinations/am/transform.js:301]: done

**Review by koladilip — DISMISSED** [review]

File: src/v0/destinations/am/utils.js [hunk 5]
```diff
@@ -82,6 +82,31 @@ const getEventId = (payload, sourceKey) => {
   return undefined;
 };
 
+/**
+ * generates the unsetObject and returns it
```
Comment (koladilip) [inline src/v0/destinations/am/utils.js:86]: is there doc link in amplitude website for this?

Comment (anantjain45823) [reply src/v0/destinations/am/utils.js:86]: (https://www.docs.developers.amplitude.com/analytics/apis/http-v2-api/#keys-for-the-event-argument:~:text=exceed%2040%20layers.-,user_properties,-Optional.%20Object.%20A)

File: src/v0/destinations/am/utils.js [hunk 6]
```diff
@@ -82,6 +82,31 @@ const getEventId = (payload, sourceKey) => {
...
+  };
+  return unsetObj = {
+        "Unset1": "-",
+        "Unset2": "-"
+      }
+ */
+const getUnsetObj = (message) => {
+  const fieldsToUnset = get(message, 'integrations.Amplitude.fieldsToUnset');
+  let unsetObject;
+  if (fieldsToUnset) {
```
Comment (sanpj2292) [inline src/v0/destinations/am/utils.js:104]: Let's do `Array.isArray()` here ?

File: src/v0/destinations/am/utils.js (hunk 5 above)
Comment (sanpj2292) [reply src/v0/destinations/am/utils.js:86]: Let's add reference documentation in source-code itself ?

Comment (anantjain45823) [reply src/v0/destinations/am/utils.js:86]: Done

File: src/v0/destinations/am/utils.js (hunk 6 above)
Comment (anantjain45823) [reply src/v0/destinations/am/utils.js:104]: done

File: src/v0/destinations/am/utils.js [hunk 7]
```diff
@@ -82,6 +82,32 @@ const getEventId = (payload, sourceKey) => {
...
+  return unsetObj = {
+        "Unset1": "-",
+        "Unset2": "-"
+      }
+  AM docs: https://www.docs.developers.amplitude.com/analytics/apis/http-v2-api/#keys-for-the-event-argument:~:text=exceed%2040%20layers.-,user_properties,-Optional.%20Object.%20A
+ */
+const getUnsetObj = (message) => {
+  const fieldsToUnset = get(message, 'integrations.Amplitude.fieldsToUnset');
+  let unsetObject;
+  if (fieldsToUnset && Array.isArray(fieldsToUnset)) {
```
Comment (sanpj2292) [inline src/v0/destinations/am/utils.js:105]: Just `Array.isArray(fieldsToUnset)` should suffice

Comment (sanpj2292) [reply src/v0/destinations/am/utils.js:105]: Can you once confirm ?

Comment (anantjain45823) [reply src/v0/destinations/am/utils.js]: done

**Review by yashasvibajpai — APPROVED** [review]

**Review by koladilip — APPROVED** [review]

**Review by sanpj2292 — APPROVED** [review]

**Pull request merged**